tdrs-backend/tdpservice/reports/models.py

```diff
@@ -0,0 +1,116 @@
+"""Define report models."""
+
+from django.db import models
+from django.db.models import Max
+
+from ..stts.models import STT
+from ..users.models import User
+
+
+# The Report File model was starting to explode, and I think that keeping this logic
+# in its own abstract class is better for documentation purposes.
+class File(models.Model):
+    """Abstract type representing a file stored in S3."""
+
+    class Meta:
+        """Metadata."""
+
+        abstract = True
+
+    original_filename = models.CharField(max_length=256,
+                                         blank=False,
+                                         null=False)
+
+    slug = models.CharField(max_length=256, blank=False, null=False)
+    extension = models.CharField(max_length=8, default="txt")
+
+
+class ReportFile(File):
+    """Represents a version of a report file."""
+
+    class Section(models.TextChoices):
+        """Enum for report section."""
+
+        ACTIVE_CASE_DATA = "Active Case Data"
+        CLOSE_CASE_DATA = "Close Case Data"
+        AGGREGATE_DATA = "Aggregate Data"
+        STRATUM_DATA = "Stratum Data"
+
+    class Quarter(models.TextChoices):
+        """Enum for report Quarter."""
+
+        Q1 = "Q1"
+        Q2 = "Q2"
+        Q3 = "Q3"
+        Q4 = "Q4"
+
+    class Meta:
+        """Metadata."""
+
+        constraints = [
+            models.UniqueConstraint(
+                fields=("section", "version", "quarter", "year", "stt"),
+                name="constraint_name",
+            )
+        ]
+
+    quarter = models.CharField(max_length=16,
+                               blank=False,
+                               null=False,
+                               choices=Quarter.choices)
+    year = models.CharField(max_length=16, blank=False, null=False)
+    section = models.CharField(max_length=32,
+                               blank=False,
+                               null=False,
+                               choices=Section.choices)
+
+    version = models.IntegerField()
+
+    user = models.ForeignKey(User,
+                             on_delete=models.CASCADE,
+                             related_name="user",
+                             blank=False,
+                             null=False)
+    stt = models.ForeignKey(STT,
+                            on_delete=models.CASCADE,
+                            related_name="sttRef",
+                            blank=False,
+                            null=False)
+
+    @classmethod
+    def create_new_version(self, data):
```

Comment:
This looks like a solid approach! A potential refactor could use a timestamp rather than implementing our own incrementation logic - that way we could grab the most recent version based on the entry with the most recent timestamp.

Reply:
I am gonna have to think about that one, cause your right it could be done with a time stamp.

Reply:
@hursey013 @riatzukiza My thought on this was that sometime down the road we would probably need some kind of endpoint that would list all versions so they could be accessed for auditing purposes. Having a version number would make that list easier to read. I do think that adding a timestamp is a good idea as well.